### PR TITLE
DRAFT: refactor: introduce new Scaleway SDK

### PR DESCRIPTION
**Pull request by QuentinBrosse**

### CODE REVIEW

File: scaleway/config.go [hunk 1]
```diff
@@ -3,36 +3,175 @@ package scaleway
...
+	err = fetchServerAvailabilities(client)
+	if err != nil {
+		log.Printf("error: cannot fetch server availabilities: %s", err)
+	}
+
+	return meta, nil
 }
 
-// client is a bridge between sdk.HTTPClient interface and retryablehttp.Client
+var userAgent = fmt.Sprintf("terraform-provider/%s;terraform/%s", version, terraform.VersionString())
```
Comment (kindermoumoute) [inline scaleway/config.go]: Is this user agent format documented somewhere?

File: scaleway/config.go [hunk 2]
```diff
@@ -11,28 +13,128 @@ import (
...
+	c.RetryMax = 3
+	c.RetryWaitMax = 2 * time.Minute
+	c.Logger = log.New(os.Stderr, "", 0)
+	c.RetryWaitMin = time.Minute
+	c.CheckRetry = func(_ context.Context, resp *http.Response, err error) (bool, error) {
+		if resp == nil {
+			return true, err
+		}
+		if resp.StatusCode == http.StatusTooManyRequests {
+			return true, err
```
Comment (kindermoumoute) [inline scaleway/config.go]: `resp == nil || resp.StatusCode == http.StatusTooManyRequests`

File: scaleway/provider.go [hunk 3]
```diff
@@ -1,75 +1,152 @@
...
 // Provider returns a terraform.ResourceProvider.
 func Provider() terraform.ResourceProvider {
+
+	// Init the SDK logger.
+	scwLogger.SetLogger(logger{})
+
+	// Init the Scaleway config.
+	scwConfig, err := scwconfig.Load()
+	if err != nil {
+		log.Fatalf("error: cannot load configuration: %s", err)
```
Comment (kindermoumoute) [inline scaleway/provider.go]: remove `error:`

File: scaleway/config.go [hunk 4]
```diff
@@ -11,28 +13,128 @@ import (
...
+	client, err := c.GetScwClient()
+	if err != nil {
+		return nil, err
+	}
+	meta.scwClient = client
+
+	// Deprecated Scaleway Client
+	deprecatedClient, err := c.GetDeprecatedClient()
+	if err != nil {
+		return nil, fmt.Errorf("error: cannot create deprecated client: %s", err)
```
Comment (kindermoumoute) [inline scaleway/config.go]: remove `error:`

File: scaleway/config.go [hunk 5]
```diff
@@ -11,28 +13,128 @@ import (
...
+		options = append(options, scw.WithDefaultRegion(c.DefaultRegion))
+	}
+
+	if c.DefaultZone != "" {
+		options = append(options, scw.WithDefaultZone(c.DefaultZone))
+	}
+
+	client, err := scw.NewClient(options...)
+	if err != nil {
+		return nil, fmt.Errorf("error: cannot create SDK client: %s", err)
```
Comment (kindermoumoute) [inline scaleway/config.go]: remove `error:`

File: scaleway/provider.go [hunk 6]
```diff
@@ -1,75 +1,152 @@
...
 				Type:        schema.TypeString,
 				Optional:    true,
-				DefaultFunc: schema.EnvDefaultFunc("SCALEWAY_ACCESS_KEY", nil),
-				Deprecated:  "Use `token` instead.",
-				Description: "The API key for Scaleway API operations.",
+				Description: "The Scaleway access key.",
+				DefaultFunc: schema.SchemaDefaultFunc(func() (interface{}, error) {
+					// Keep the deprecated behavior
+					if accessKey := os.Getenv("SCALEWAY_ACCESS_KEY"); accessKey != "" {
+						log.Printf("[WARN] SCALEWAY_ACCESS_KEY is deprecated, please use SCW_ACCESS_KEY instead")
```
Comment (kindermoumoute) [inline scaleway/provider.go]: Use the global logger `logger.Warnf()`

File: scaleway/provider.go [hunk 7]
```diff
@@ -1,75 +1,152 @@
...
+			},
+			"region": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: "The Scaleway default region to use for your resources.",
+				DefaultFunc: schema.SchemaDefaultFunc(func() (interface{}, error) {
+					// Keep the deprecated behavior
+					// Note: The deprecated region format conversion is handled in `config.GetDeprecatedClient`.
+					if region := os.Getenv("SCALEWAY_REGION"); region != "" {
+						log.Printf("[WARN] SCALEWAY_REGION is deprecated, please use SCW_DEFAULT_REGION instead")
```
Comment (kindermoumoute) [inline scaleway/provider.go]: use the global logger

File: scaleway/provider.go [hunk 8]
```diff
@@ -1,75 +1,152 @@
...
+						log.Printf("[WARN] SCALEWAY_ORGANIZATION is deprecated, please use SCW_DEFAULT_ORGANIZATION_ID instead")
+						return organization, nil
 					}
 					if path, err := homedir.Expand("~/.scwrc"); err == nil {
-						_, scwOrganization, err := readScalewayConfig(path)
+						_, scwOrganization, err := readDeprecatedScalewayConfig(path)
 						if err != nil {
-							return nil, err
+							// No error is returned here to allow user to use `organization_id`.
+							log.Printf("[ERROR] cannot parse deprecated config file: %s", err)
```
Comment (kindermoumoute) [inline scaleway/provider.go]: use the global logger

File: scaleway/config.go (hunk 1 above)
Comment (jerome-quere) [reply scaleway/config.go]: we should use space as separator `terraform-provider/%s terraform/%s`

**Pull request closed without merging**